spring-cloud-sleuth-core/src/main/java/org/springframework/cloud/sleuth/annotation/SleuthAdvisorConfig.java

```diff
@@ -183,44 +181,141 @@ public Object invoke(MethodInvocation invocation) throws Throwable {
 		try (Tracer.SpanInScope ws = tracer().withSpanInScope(span)) {
-			if (hasLog) {
-				logEvent(span, log + ".before");
-			}
-			spanTagAnnotationHandler().addAnnotatedParameters(invocation);
-			addTags(invocation, span);
+			before(invocation, span, log, hasLog);
 			return invocation.proceed();
 		} catch (Exception e) {
-			if (logger.isDebugEnabled()) {
-				logger.debug("Exception occurred while trying to continue the pointcut", e);
-			}
-			if (hasLog) {
-				logEvent(span, log + ".afterFailure");
-			}
-			span.error(e);
+			onFailure(span, log, hasLog, e);
 			throw e;
 		} finally {
-			if (hasLog) {
-				logEvent(span, log + ".after");
+			after(span, newSpan != null, log, hasLog);
+		}
+	}
+
+	private Object proceedUnderSpanReactively(
+			MethodInvocation invocation, NewSpan newSpan, ContinueSpan continueSpan) throws Throwable{
+		boolean isNewSpan = newSpan != null;
+		Span spanPrevious = tracer().currentSpan();
+		Span span;
+		if (isNewSpan || spanPrevious == null) {
+			span = tracer().nextSpan();
+			newSpanParser().parse(invocation, newSpan, span);
+		} else {
+			span = spanPrevious;
+		}
+
+		String log = log(continueSpan);
+		boolean hasLog = StringUtils.hasText(log);
+
+		try(Tracer.SpanInScope ws = tracer().withSpanInScope(span)) {
+
+			Publisher<?> publisher = (Publisher) invocation.proceed();
+
+			Mono<Span> startSpan = Mono.defer(() -> {
+				try(Tracer.SpanInScope ws1 = tracer().withSpanInScope(span)) {
+					Span spanStarted;
+					if (isNewSpan || spanPrevious == null) {
+						spanStarted = span.start();
+					} else {
+						spanStarted = span;
+					}
+
+					before(invocation, spanStarted, log, hasLog);
+					return Mono.just(spanStarted);
+				}
+			});
+
+			if(publisher instanceof Mono){
+				return startSpan.flatMap(spanStarted -> ((Mono<?>)publisher)
+						.doOnError(throwable -> {
+							try(Tracer.SpanInScope ws1 = tracer().withSpanInScope(spanStarted)) {
+								onFailure(spanStarted, log, hasLog, throwable);
+							}
+						})
+						.doOnTerminate(() -> {
+							try(Tracer.SpanInScope ws1 = tracer().withSpanInScope(spanStarted)) {
+								after(spanStarted, isNewSpan, log, hasLog);
+							}
+						}));
 			}
-			if (newSpan != null) {
-				span.finish();
+			else if(publisher instanceof Flux){
+				return startSpan.flatMapMany(spanStarted -> ((Flux<?>)publisher)
+						.doOnError(throwable -> {
+							try(Tracer.SpanInScope ws1 = tracer().withSpanInScope(spanStarted)) {
+								onFailure(spanStarted, log, hasLog, throwable);
+							}
+						})
+						.doOnTerminate(() -> {
+							try(Tracer.SpanInScope ws1 = tracer().withSpanInScope(spanStarted)) {
+								after(spanStarted, isNewSpan, log, hasLog);
+							}
+						}));
 			}
+			else {
+				throw new IllegalArgumentException("Unexpected type of publisher: "+publisher.getClass());
+			}
+		}
+	}
+
+	private void before(MethodInvocation invocation, Span span, String log, boolean hasLog) {
+		if (hasLog) {
```

Comment:
these annotations will be costly to index, are we sure we want them? Also these should be named "event" or similar, not log

Reply:
Not clear for me as it's just safe refactoring(extracted code to method to not duplicate it). All code and variable names are kept unchanged.